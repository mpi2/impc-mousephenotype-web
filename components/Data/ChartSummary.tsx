import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faArrowLeftLong } from "@fortawesome/free-solid-svg-icons";
import { Col, Modal, Row, Table } from "react-bootstrap";
import Card from "@/components/Card";
import { useRouter } from "next/router";
import { formatAlleleSymbol } from "@/utils";
import { PropsWithChildren, ReactNode, useState } from "react";
import Link from "next/link";


type ChartSummaryProps = {
  datasetSummary: any;
  additionalContent?: ReactNode,
  title?: string;
}
const ChartSummary = ({ title, additionalContent = null, datasetSummary, children }: PropsWithChildren<ChartSummaryProps>) => {
  const router = useRouter();
  const [showMetadataModal, setShowMetadataModal ] = useState(false);
  const allele = formatAlleleSymbol(datasetSummary["alleleSymbol"]);
  const totalMice = Object.keys(datasetSummary["summaryStatistics"]).reduce(
    (acc, key) => {
      return (
        acc +
        (key.includes("Count") ? datasetSummary["summaryStatistics"][key] : 0)
      );
    },
    0
  );

  const metadataArray = datasetSummary.metadataValues?.[0].split('|') || [];

  return (
    <Card>
      <div>
        <button
          style={{ border: 0, background: "none", padding: 0 }}
          onClick={() => {
            router.back();
          }}
        >
          <a href="#" className="grey mb-3 small">
            <FontAwesomeIcon icon={faArrowLeftLong} /> GO BACK
          </a>
        </button>
        <h1>
          <strong>{!!title ? title : datasetSummary["geneSymbol"] + ' data charts'}</strong>
        </h1>
        {additionalContent}
      </div>
      <h2>Description of the experiments performed</h2>
      <Row>
        <Col md={7} style={{ borderRight: "1px solid #ddd" }}>
          {children ? children : (
            <>
              <p>
                A <strong>{datasetSummary["procedureName"]}</strong> phenotypic
                assay was performed on {totalMice} mice. The charts show the
                results of measuring{" "}
                <strong>{datasetSummary["parameterName"]}</strong> in{" "}
                {datasetSummary["summaryStatistics"]["femaleMutantCount"]} female,{" "}
                {datasetSummary["summaryStatistics"]["maleMutantCount"]} male
                mutants compared to{" "}
                {datasetSummary["summaryStatistics"]["femaleControlCount"]}{" "}
                female, {datasetSummary["summaryStatistics"]["maleControlCount"]}{" "}
                male controls. The mutants are for the {allele[0]}
                <sup>{allele[1]}</sup> allele.
              </p>
              <p className="small">
                * The high throughput nature of the IMPC means that large control
                sample sizes may accumulate over a long period of time. See the
                animal welfare guidelines for more information.
              </p>
            </>
          )}
        </Col>
        <Col md={5} className="small">
          <p className="mb-2">
              <span style={{ display: "inline-block", width: 180 }}>
                Zygosity
              </span>
            <strong>{datasetSummary["zygosity"]}</strong>
          </p>
          <p className="mb-2">
              <span style={{ display: "inline-block", width: 180 }}>
                Testing protocol
              </span>
            <strong>
              <Link
                className="link primary"
                href={`https://www.mousephenotype.org/impress/ProcedureInfo?action=list&procID=${datasetSummary.procedureStableKey}&pipeID=${datasetSummary.pipelineStableKey}`}
              >
                {datasetSummary["procedureName"]}
              </Link>
            </strong>
          </p>
          <p className="mb-2">
              <span style={{ display: "inline-block", width: 180 }}>
                Testing environment
              </span>
            <strong className="primary link" onClick={() => setShowMetadataModal(true)}>
              Lab conditions and equipment
            </strong>
          </p>
          <p className="mb-2">
              <span style={{ display: "inline-block", width: 180 }}>
                Measured value
              </span>
            <strong>{datasetSummary["parameterName"]}</strong>
          </p>
          <p className="mb-2">
              <span style={{ display: "inline-block", width: 180 }}>
                Life stage
              </span>
            <strong>{datasetSummary["lifeStageName"]}</strong>
          </p>
          <p className="mb-2">
              <span style={{ display: "inline-block", width: 180 }}>
                Background Strain
              </span>
            <strong>{datasetSummary["geneticBackground"]}</strong>
          </p>
          <p className="mb-2">
              <span style={{ display: "inline-block", width: 180 }}>
                Phenotyping center
              </span>
            <strong>{datasetSummary["phenotypingCentre"]}</strong>
          </p>
          <p className="mb-2">
              <span style={{ display: "inline-block", width: 180 }}>
                Associated Phenotype
              </span>
            {!!datasetSummary["significantPhenotype"]?.["id"] ? (
              <Link href={`/phenotypes/${datasetSummary["significantPhenotype"]["id"]}`}>
                <strong className="link primary">{datasetSummary["significantPhenotype"]?.["name"]}</strong>
              </Link>
            ) : (
              <strong>No significant association</strong>
            )}
          </p>
        </Col>
      </Row>
      <Modal show={showMetadataModal} onHide={() => setShowMetadataModal(false)}>
        <Modal.Header closeButton>
          <Modal.Title>Experimental conditions</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <Table striped borderless>
            <tbody>
            {metadataArray.map(item => item.split('=')).map(([label, value]) =>
              <tr>
                <td>{label}</td>
                <td>{value}</td>
              </tr>
            )}
            </tbody>
          </Table>
        </Modal.Body>
      </Modal>
    </Card>
  )
};

export default ChartSummary;