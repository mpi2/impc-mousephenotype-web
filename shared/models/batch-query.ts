export type GoTerm = {
  aspect: string;
  assigned_by: string;
  db_names: string;
  db_type: string;
  evidence_code: string;
  go_id: string;
  go_name: string;
  go_term_specificity: number;
  references: string;
};

export type BatchQueryItem = {
  geneId: string;
  mouseGeneSymbol: string;
  humanGeneIds: string | Array<string>;
  humanGeneSymbols: string | Array<string>;
  allSignificantSystems: Array<string>;
  allSignificantPhenotypes: Array<string>;
  alleles: Array<{
    allele: string;
    significantPhenotypes: Array<string>;
    significantLifeStages: Array<string>;
    significantSystems: Array<string>;
    notSignificantPhenotypes: Array<string>;
    notSignificantSystems: Array<string>;
  }>;
  goTerms: Array<GoTerm>;
};

export type SelectedAlleleData = {
  alelleSymbol: string;
  phenotypes: Array<string>;
};
