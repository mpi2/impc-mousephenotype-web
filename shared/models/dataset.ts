
type SummaryStatistics = {
  bothMutantCount:    number;
  bothMutantMean:     number;
  bothMutantSd:       number;
  femaleControlCount: number;
  femaleControlMean:  number;
  femaleControlSd:    number;
  femaleMutantCount:  number;
  femaleMutantMean:   number;
  femaleMutantSd:     number;
  maleControlCount:   number;
  maleControlMean:    number;
  maleControlSd:      number;
  maleMutantCount:    number;
  maleMutantMean:     number;
  maleMutantSd:       number;
};
type StatisticalMethod = {
  method: string;
  attributes: {
    femaleEffectSize:                number | null;
    femaleKoEffectPValue:            number;
    femaleKoEffectStderrEstimate:    number;
    femaleKoParameterEstimate:       number;
    femalePercentageChange:          number | null;
    maleKoEffectPValue:              number;
    maleKoEffectStderrEstimate:      number;
    maleKoParameterEstimate:         number;
    malePercentageChange:            number | null;
    genotypeEffectPValue:            number;
    genotypeEffectStderrEstimate:    number;
    group1Genotype:                  string;
    group1ResidualsNormalityTest:    number;
    group2Genotype:                  string;
    group2ResidualsNormalityTest:    number;
    interactionEffectPValue:         number | null;
    interactionSignificant:          boolean | null;
    interceptEstimate:               number;
    interceptEstimateStderrEstimate: number;
    maleEffectSize:                  number | null;
    sexEffectPValue:                 number | null;
    sexEffectParameterEstimate:      number | null;
    sexEffectStderrEstimate:         number | null;
    batchSignificant:                boolean;
    varianceSignificant:             boolean;
  }
}

export type Dataset = {
  id:                     string;
  alleleAccessionId:      string;
  alleleName:             string;
  alleleSymbol:           string;
  colonyId:               string;
  dataType:               string;
  geneSymbol:             string;
  geneticBackground:      string;
  intermediatePhenotypes: Array<{ name: string, id: string }>;
  lifeStageAcc:           string;
  lifeStageName:          string;
  metadataGroup:          string;
  metadataValues:         Array<string>;
  mgiGeneAccessionId:     string;
  parameterName:          string;
  parameterStableId:      string;
  parameterStableKey:     number;
  phenotypeSex:           Array<string>;
  phenotypingCentre:      string;
  pipelineName:           string;
  pipelineStableId:       string;
  pipelineStableKey:      number;
  procedureGroup:         string;
  procedureName:          string;
  procedureStableId:      string;
  procedureStableKey:     number;
  productionCentre:       string;
  projectName:            string;
  reportedEffectSize:     number;
  reportedPValue:         number;
  resourceFullName:       string;
  resourceName:           string;
  sex:                    string;
  zygosity:               string;
  significant:            boolean;
  significantPhenotype:   { name: string, id: string };
  datasetId:              string;
  status:                 string;
  strainAccessionId:      string;
  strainName:             string;
  summaryStatistics:      SummaryStatistics;
  statisticalMethod:      StatisticalMethod;
  topLevelPhenotypes:     Array<{ name: string, id: string }>;
}