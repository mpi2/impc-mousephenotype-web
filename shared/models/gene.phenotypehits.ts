
export type GenePhenotypeHits = {
  alleleAccessionId: string;
  alleleName: string;
  alleleSymbol: string;
  datasetId: string;
  effectSize: number;
  intermediatePhenotypes: Array<{ id: string; name: string; }>;
  lifeStageName: string;
  mgiGeneAccessionId: string;
  pValue: number;
  parameterName: string;
  parameterSatableId: string;
  phenotype: { id: string; name: string; };
  phenotypingCentre: string;
  pipelineStableId: string;
  procedureName: string;
  procedureStableId: string;
  projectName: string;
  sex: string;
  topLevelPhenotypes: Array<{ id: string; name: string; }>;
  zygosity: string;
}