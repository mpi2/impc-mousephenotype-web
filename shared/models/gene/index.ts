export type { GeneSearchResponse, GeneSearchResponseItem } from './search-response';
export type { GeneSummary } from './summary';
export type { GenePhenotypeHits } from './phenotype-hits';