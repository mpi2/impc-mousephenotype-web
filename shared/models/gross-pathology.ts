
export type GrossPathologyDataset = {
  alleleAccessionId: string;
  alleleSymbol: string;
  dataPoint: null;
  externalSampleId: string;
  lifeStageName: string;
  ontologyTerms: Array<{ termId: string; termName: string }>;
  parameterName: string;
  phenotypingCenter: string;
  pipelineStableId: string;
  procedureStableId: string;
  textValue: null;
  zygosity: string;
};

export type GrossPathology = {
  id: string;
  mgiGeneAccessionId: string;
  parameterStableId: string;
  datasets: Array<GrossPathologyDataset>
}