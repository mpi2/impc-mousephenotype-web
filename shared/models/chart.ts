import { Dataset } from "@/models/dataset";

export type GeneralChartProps = {
  datasetSummary: Dataset;
  isVisible: boolean;
};