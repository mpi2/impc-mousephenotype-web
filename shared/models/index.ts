import * as Gene from './gene';
import * as Phenotype from './phenotype';
import type { PhenotypeRef } from './phenotype-ref';
import { MetabolismGeneData } from "@/pages/metabolism";
import { HistopathologyResponse, Histopathology } from './histopathology';
import { Dataset } from './dataset';

export type { PhenotypeRef } from './phenotype-ref';
export type { TableCellProps } from './TableCell';
export type { HistopathologyResponse, Histopathology } from './histopathology';
export type { Dataset } from './dataset';
export type { Gene };
export type { Phenotype };

type Model =
  PhenotypeRef | typeof Gene | typeof Phenotype | MetabolismGeneData | Dataset
  | HistopathologyResponse | Histopathology;
export type { Model };