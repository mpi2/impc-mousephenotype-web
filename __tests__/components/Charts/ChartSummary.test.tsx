import { render, screen } from '@testing-library/react';
import ChartSummary from "@/components/Data/ChartSummary/ChartSummary";
import { Dataset } from "@/models";
import userEvent from "@testing-library/user-event";

const datasetSummary: Dataset = {
  "id": "66617d66820fc2a6e99d04c9",
  "alleleAccessionId": "MGI:6342258",
  "alleleName": "endonuclease-mediated mutation 1, Jackson",
  "alleleSymbol": "Otog<em1(IMPC)J>",
  "colonyId": "JR34077",
  "dataType": "categorical",
  "geneSymbol": "Otog",
  "geneticBackground": "involves: C57BL/6NJ",
  "intermediatePhenotypes": [{ id: "MP:0010678", name: "abnormal skin adnexa morphology" }],
  "lifeStageAcc": "IMPCLS:0005",
  "lifeStageName": "Early adult",
  "metadataGroup": "ba6c7cda9f0d4ce7d9a676c2aef86e22",
  "metadataValues": [
    "Experimenter ID = 136|Location of test = Open bench|Number of animals in cage = 3|Size of squares in arena = 68.89",
    "Experimenter ID = 136|Location of test = Open bench|Number of animals in cage = 5|Size of squares in arena = 68.89"
  ],
  "mgiGeneAccessionId": "MGI:1202064",
  "parameterName": "Vibrissae - appearance",
  "parameterStableId": "IMPC_CSD_024_001",
  "parameterStableKey": 1804,
  "phenotypeSex": ["male", "female"],
  "phenotypingCentre": "JAX",
  "pipelineName": "JAX Pipeline",
  "pipelineStableId": "JAX_001",
  "pipelineStableKey": 12,
  "procedureGroup": "IMPC_CSD",
  "procedureName": "Combined SHIRPA and Dysmorphology",
  "procedureStableId": "IMPC_CSD_001",
  "procedureStableKey": 82,
  "productionCentre": "JAX",
  "projectName": "JAX",
  "reportedEffectSize": 0.370222929936306,
  "reportedPValue": 0.0000121024575839839,
  "resourceFullName": "IMPC",
  "resourceName": "IMPC",
  "sex": "female",
  "zygosity": "homozygote",
  "significant": true,
  "significantPhenotype": {
    "name": "abnormal vibrissa morphology",
    "id": "MP:0002098"
  },
  "statisticalMethod": {
    "method": "Fisher Exact Test framework",
    "attributes": {
      "femaleEffectSize": null,
      "femaleKoEffectPValue": 0.0000106734229163214,
      "femaleKoEffectStderrEstimate": null,
      "femaleKoParameterEstimate": null,
      "femalePercentageChange": null,
      "maleKoEffectPValue": 1,
      "maleKoEffectStderrEstimate": null,
      "maleKoParameterEstimate": null,
      "malePercentageChange": null,
      "genotypeEffectPValue": null,
      "genotypeEffectStderrEstimate": null,
      "group1Genotype": null,
      "group1ResidualsNormalityTest": null,
      "group2Genotype": null,
      "group2ResidualsNormalityTest": null,
      "interactionEffectPValue": null,
      "interactionSignificant": null,
      "interceptEstimate": null,
      "interceptEstimateStderrEstimate": null,
      "maleEffectSize": null,
      "sexEffectPValue": null,
      "sexEffectParameterEstimate": null,
      "sexEffectStderrEstimate": null,
      "batchSignificant": null,
      "varianceSignificant": null
    }
  },
  "datasetId": "5a124934293cf296449781e17558d227",
  "status": "Successful",
  "strainAccessionId": "MGI:3056279",
  "strainName": "C57BL/6NJ",
  "summaryStatistics": {
    "bothMutantCount": null,
    "bothMutantMean": null,
    "bothMutantSd": null,
    "femaleControlCount": 1884,
    "femaleControlMean": null,
    "femaleControlSd": null,
    "femaleMutantCount": 5,
    "femaleMutantMean": null,
    "femaleMutantSd": null,
    "maleControlCount": 1884,
    "maleControlMean": null,
    "maleControlSd": null,
    "maleMutantCount": 3,
    "maleMutantMean": null,
    "maleMutantSd": null
  },
  "topLevelPhenotypes": [
    {
      "name": "integument phenotype",
      "id": "MP:0010771"
    }
  ]
}


describe('Chart summary component', () => {
  it('renders correctly', () => {
    const {container} = render(
      <ChartSummary
        datasetSummary={datasetSummary}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('display metadata model on link click', async () => {
    const user = userEvent.setup();
    render(<ChartSummary datasetSummary={datasetSummary} />);
    await user.click(await screen.findByText(/Lab conditions and equipment/i));
    expect(await screen.findByRole('dialog')).toBeInTheDocument();
    await user.click(await screen.findByRole("button"));
    expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
  });
});