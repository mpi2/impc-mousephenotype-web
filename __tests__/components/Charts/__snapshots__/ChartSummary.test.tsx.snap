// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart summary component renders correctly 1`] = `
<div>
  <div
    class="card"
  >
    <div>
      <div
        class="mb-4"
        style="display: flex; align-items: center; gap: 2rem;"
      >
        <h1
          style="margin: 0px;"
        >
          <strong>
            <i>
              Otog
            </i>
             data charts
          </strong>
        </h1>
        <span
          style="color: rgb(121, 118, 118); text-transform: uppercase; font-weight: 500; line-height: 1.2;"
        >
          Vibrissae - appearance
        </span>
      </div>
      <span
        class="mb-4"
        style="display: inline-block; font-size: 1.2rem;"
      >
        Combination tested with the lowest p-value of
         
        <strong>
          1.07
          x10
          <sup>
            -5
          </sup>
        </strong>
      </span>
    </div>
    <div
      class="details"
    >
      <div
        class="label"
      >
        Experiments performed
      </div>
      <div
        class="content"
      >
        <p>
          A 
          <strong>
            Combined SHIRPA and Dysmorphology
          </strong>
           phenotypic assay was performed on 
          3776
           mice. The charts show the results of measuring
           
          <strong>
            Vibrissae - appearance
          </strong>
           in
           
          5
           
          female, 
          3
           
          male mutants compared to
           
          1884
           
          female,
           
          1884
           male controls. The mutants are for the
           
          <i>
            Otog
            <sup>
              em1(IMPC)J
            </sup>
          </i>
           
          allele.
        </p>
      </div>
      <div
        class="label"
      >
        Associated phenotype
      </div>
      <div
        class="content"
      >
        <a
          href="/phenotypes/MP:0002098"
        >
          <span
            class="link primary"
          >
            abnormal vibrissa morphology
          </span>
        </a>
      </div>
      <div
        class="label"
      >
        Zygosity
      </div>
      <div
        class="content"
      >
        <span>
          homozygote
        </span>
      </div>
      <div
        class="label"
      >
        Testing protocol
      </div>
      <div
        class="content"
      >
        <a
          class="link primary"
          href="https://www.mousephenotype.org/impress/ProcedureInfo?action=list&procID=82&pipeID=12"
        >
          Combined SHIRPA and Dysmorphology
        </a>
      </div>
      <div
        class="label"
      >
        Measured value
      </div>
      <div
        class="content"
      >
        Vibrissae - appearance
      </div>
      <div
        class="label"
      >
        Testing environment
      </div>
      <div
        class="content"
      >
        <span
          class="primary link"
        >
          Lab conditions and equipment
        </span>
      </div>
      <div
        class="label"
      >
        Life stage
      </div>
      <div
        class="content"
      >
        <span>
          Early adult
        </span>
      </div>
      <div
        class="label"
      >
        Background Strain
      </div>
      <div
        class="content"
      >
        <span>
          involves: C57BL/6NJ
        </span>
      </div>
      <div
        class="label"
      >
        Phenotyping center
      </div>
      <div
        class="content"
      >
        <span>
          JAX
        </span>
      </div>
      <div
        class="label"
      />
      <div
        class="content"
      >
        <p
          class="small"
        >
          * The high throughput nature of the IMPC means that large control sample sizes may accumulate over a long period of time. See the animal welfare guidelines for more information.
        </p>
      </div>
    </div>
  </div>
</div>
`;
